snuba/clickhouse/query_formatter.py

```diff
@@ -0,0 +1,179 @@
+from typing import Sequence, Tuple, Union, NamedTuple, Mapping, Optional
+
+from snuba import settings as snuba_settings
+from snuba.clickhouse.formatter import ClickhouseExpressionFormatter
+from snuba.clickhouse.query import Query
+from snuba.datasets.schemas import RelationalSource
+from snuba.query.composite import CompositeQuery
+from snuba.query.parsing import ParsingContext
+from snuba.request.request_settings import RequestSettings
+
+
+class FormattedQuery(NamedTuple):
+    """
+    Used to move around a query data structure after all clauses have
+    been formatted but such that it can be still serialized
+    differently for different usages (running the query or tracing).
+    """
+
+    # TODO: This is going to be a little more complex for subqueries
+    # and joins in order to have a useful tracing representation.
+    clauses: Sequence[Tuple[str, str]]
+
+    def get_sql(self, format: Optional[str] = None) -> str:
+        query = " ".join([c for _, c in self.clauses])
+        if format is not None:
+            query = f"{query} FORMAT {format}"
+
+        return query
+
+    def get_mapping(self) -> Mapping[str, str]:
+        return dict(self.clauses)
+
+
+FormattableQuery = Union[Query, CompositeQuery[RelationalSource]]
+
+
+def format_query(query: Query, settings: RequestSettings) -> FormattedQuery:
+    """
+    Replaces the AstSqlQuery abstraction.
+
+    TODO: Remove this method entirely and move the sampling logic
+    into a query processor.
+    """
+
+    if not query.get_from_clause().supports_sample():
+        sample_rate = None
+    else:
+        if query.get_sample():
+            sample_rate = query.get_sample()
+        elif settings.get_turbo():
+            sample_rate = snuba_settings.TURBO_SAMPLE_RATE
+        else:
+            sample_rate = None
+    query.set_sample(sample_rate)
+
+    return _format_query_impl(query)
+
+
+def _format_query_impl(query: FormattableQuery) -> FormattedQuery:
```

Comment:
This is pretty much the same implementation that before was in the AstQuery class. I just renamed some variables and extracted the query processing related with settings.